pkg/pub_package_reader/lib/pub_package_reader.dart

```diff
@@ -79,6 +80,12 @@ Future<PackageSummary> summarizePackageArchive(
   /// The maximum number of files in the archive.
   /// TODO: set this lower once we scan the existing archives
   int maxFileCount = 64 * 1024,
+
+  /// The timestamp when the archive was uploaded and accepted.
+  ///
+  /// If specified, some packages may get a transparent rewrite of their
+  /// assets to preserve backwards-compatible behavior on pub.dev.
+  DateTime? created,
```

Comment:
Why does this need to be optional? Can't we just patch all locations that call `summarizePackageArchive`?

Just a suggestion:
```suggestion
  /// [DateTime] when the archive was uploaded / published.
  ///
  /// Used to ensure that constraints in place when a package was published
  /// are considered when summarizing assets. Ensuring we preserve
  /// backwards-compatibility with packages that does not satisfy constraints
  /// enforced at publish-time today.
  ///
  /// Defaults to [DateTime.now], if not specified.
  DateTime? published,
```